### PR TITLE
refactor: HTTP config

### LINKED ISSUE #2277: HTTP config refactor
The HTTP config can be simplified by moving support for autotls to an external tool such as https://go-acme.github.io/lego/

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2277 

## Description

This PR is a refactor of the HTTP config and is part of the larger config refactor #2257 

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

make test

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: http/server.go [hunk 1]
```diff
@@ -13,304 +13,171 @@ package http
...
+
+// ListenAndServe listens for and serves incoming connections.
+func (s *Server) ListenAndServe() error {
+	listener, err := net.Listen("tcp", s.options.Address)
+	if err != nil {
+		return err
 	}
+	// ignore close errors as they cannot be handled
+	// from the caller of this method
+	defer listener.Close() //nolint:errcheck
```
Comment (AndrewSisley) [inline http/server.go:166]: question: Why can they not be handled but other errors appear to be so?  And is that not up to any callers of this public function to decide?  Should the choice to discard the error not be made by the caller?

File: http/server.go [hunk 2]
```diff
@@ -13,304 +13,171 @@ package http
...
-	}
-
-	// Save the address on the server in case the port was set to random
-	// and that we want to see what was assigned.
-	s.address = s.listener.Addr().String()
-
-	return nil
+// Server struct holds the Handler for the HTTP API.
+type Server struct {
+	address atomic.Value
```
Comment (AndrewSisley) [inline http/server.go:119]: question: Why is this atomic? It only appears to be set once, in `ListenAndServe`, and that function cannot be called multiple times concurrently no?

todo: If it remains atomic, please document why.

**Review by AndrewSisley — APPROVED** [review]
praise:  Code looks much simpler, thank you :)

Only a couple of comments from me, but please address them before merge :)

File: http/server.go (hunk 2 above)
Comment (nasdf) [reply http/server.go:123]: I've added documentation to clarify the use of the atomic value.

Comment (AndrewSisley) [reply http/server.go:123]: Ah, you want `AssignedAddr` to return empty if `ListenAndServe` is in progress?  Would it not be better to lock, and block `AssignedAddr` from returning until the address has been assigned? (i.e. no atomic, but a RWLock)

File: http/server.go (hunk 1 above)
Comment (nasdf) [reply http/server.go]: I've removed this `Close` call as it is redundant. After taking a closer look the `Serve` and `ServeTLS` functions will close the listener.

File: http/server.go (hunk 2 above)
Comment (nasdf) [reply http/server.go]: I agree that would be better, but I realized we don't use this method anywhere outside of `server_test.go`. I've removed the function. 

As a side note: tests that need a random port assignment should use the `httptest` package instead.

Comment (fredcarle) [reply http/server.go]: > As a side note: tests that need a random port assignment should use the httptest package instead.

That's a really nice change. Thanks for removing the need to keep track of the randomly assigned port on the http server :)

**Pull request merged**